source/link/linker.cpp

```diff
@@ -16,1183 +16,277 @@
-                                 std::vector<std::vector<uint32_t>> &binaries,
-                                 size_t num_binaries, uint32_t max_index,
-                                 uint32_t version,
-                                 spv_binary_t& merged_spvbinary,
-                                 bool create_library) {
-  spv_position_t position = {};
 
-  std::vector<uint32_t> indices(num_binaries, 5u);
-
-  std::unordered_map<spv::Id, const char *> exports;
-  std::unordered_map<spv::Id, const char *> imports;
-  std::unordered_map<spv::Id, spv::Id> link_table;
-  std::vector<spv::Id> builtins;
-
-  uint32_t* merged_binary = merged_spvbinary.code;
-  size_t& merged_length = merged_spvbinary.wordCount;
-
-  // Generate new header
-  SetWord(merged_binary, 0u, spv::MagicNumber);
-  SetWord(merged_binary, 1u, version);
-  SetWord(merged_binary, 2u, 0u);
-  SetWord(merged_binary, 3u, max_index);
-  SetWord(merged_binary, 4u, 0u);
-  merged_length = 5u;
-
-  const auto CopySections = [&binaries,&indices,num_binaries,
-                             &merged_binary,&merged_length](const std::function<bool (spv::Op)>& check_section) {
-    for (uint32_t i = 0u; i < num_binaries; ++i) {
-      const uint32_t* const binary = binaries[i].data();
-      const size_t length = binaries[i].size();
-      const uint32_t index = indices[i];
-      uint32_t j = index;
-      uint32_t prev_j = j;
-
-      while (j < length) {
-        uint32_t word_count;
-        const spv::Op opcode = ParseOpcode(binary, j, word_count);
-
-        prev_j = j;
-
-        if (!check_section(opcode))
-          break;
-
-        j += word_count;
+  for (const auto& module : modules)
+    for (const auto& insn : module->capabilities())
+      linkedModule->AddCapability(MakeUnique<Instruction>(insn));
+
+  for (const auto& module : modules)
+    for (const auto& insn : module->extensions())
+      linkedModule->AddExtension(MakeUnique<Instruction>(insn));
+
+  for (const auto& module : modules)
+    for (const auto& insn : module->ext_inst_imports())
+      linkedModule->AddExtInstImport(MakeUnique<Instruction>(insn));
+
+  // FIXME
+  for (const auto& module : modules)
+    linkedModule->SetMemoryModel(MakeUnique<Instruction>(*module->GetMemoryModel()));
+
+  for (const auto& module : modules)
+    for (const auto& insn : module->entry_points())
+      linkedModule->AddEntryPoint(MakeUnique<Instruction>(insn));
+
+  for (const auto& module : modules)
+    for (const auto& insn : module->execution_modes())
+      linkedModule->AddExecutionMode(MakeUnique<Instruction>(insn));
+
+  for (const auto& module : modules)
+    for (const auto& insn : module->debugs())
+      linkedModule->AddDebugInst(MakeUnique<Instruction>(insn));
+
+  for (const auto& module : modules)
+    for (const auto& insn : module->annotations())
+      linkedModule->AddAnnotationInst(MakeUnique<Instruction>(insn));
+
+  for (const auto& module : modules)
```

Comment:
Add TODO: Must make unique non-aggregate non-pointer types.  (Pointer types now can be non-unique in SPV_KHR_variable_pointers)